vermouth/processors/do_links.py

```diff
@@ -261,36 +262,147 @@ def _build_link_interaction_from(molecule, interaction, match):
     )
     return new_interaction
 
-
 class DoLinks(Processor):
     """
     Apply Links, taken from a molecule's force field, to the molecule.
     """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.applied_links = defaultdict(dict)
+        self.current_match = None
+        self.current_link = None
+        self.nodes_to_remove = []
+
+    def replace_attributes(self, molecule, match):
+        """
+        Given any nodes in `self.current_link` have the 'replace'
+        attribute, replace the attribute in molecule with the new
+        value. If the atomname is set to None stage the node to be
+        removed.
+
+        Parameters
+        ----------
+        molecule: :class:`vermouth.Molecule`
+        match: dict
+            dict relating nodes in `self.current_link` to those in molecule
+        """
+        for node, node_attrs in self.current_link.nodes.items():
+            if 'replace' in node_attrs:
+                if node_attrs['replace'].get('atomname', False) is None:
+                    self.nodes_to_remove.append(match[node])
+                else:
+                    node_mol = molecule.nodes[match[node]]
+                    node_mol.update(node_attrs['replace'])
+        return molecule
+
+    def add_interactions(self, molecule, match):
+        """
+        Add interactions from link to the `self.applied_interactions` dict.
+        We temporarily store them in this dict since links can remove and or
+        replace interactions but the built in replace_interaction function in
+        molecule requires looping over the complete interactions dict, which
+        is slow. Having interactions in a dict with key also takes care of all
+        symmetry issues.
+
+        Parameters
+        ----------
+        molecule: :class:`vermouth.Molecule`
+        match: dict
+            dict relating nodes in `self.current_link` to those in molecule
+        """
+        for inter_type in self.current_link.interactions:
+            for interaction in self.current_link.interactions[inter_type]:
+                new_interaction = _build_link_interaction_from(molecule, interaction, match)
+                # it is not guaranteed that interaction.atoms is a tuple
+                # the key is the atoms involved in the interaction and the version type so
+                # that multiple versions are kept and not overwritten
+                interaction_key = tuple(new_interaction.atoms) +\
+                                  tuple([new_interaction.meta.get("version", 0)])
+
+                interaction_key_rev = tuple(new_interaction.atoms[::-1]) +\
+                                      tuple([new_interaction.meta.get("version", 0)])
+
+                if interaction_key_rev in self.applied_links[inter_type]:
```

Comment:
potentially yes; I'm still thinking on the best solution. This would only become a problem for the following case:
```
[ virtualsitesn ]
1 2 3 4 -- 1
4 3 2 1 -- 1
```
However, I think that is prohibited in GROMACS because you cannot have two virtual sites in such a dependency pattern.